corehq/motech/fhir/tasks.py

```diff
@@ -394,5 +403,273 @@ def create_parent_indices(
+
+    entries = response_json['entry']
+    for entry in entries:
+        appointments.append(entry)
+
+    return appointments
+
+
+def convert_date_and_time_to_utc_timestamp(date, time):
+    date_time = date + "T" + time
+    utc_zone = dateutil.tz.gettz('UTC')
+    # Hardcoded for MGH study
+    local_zone = dateutil.tz.gettz('America/New_York')
+    local_datetime = datetime.fromisoformat(date_time)
+    local_datetime = local_datetime.replace(tzinfo=local_zone)
+    utc_datetime = local_datetime.astimezone(utc_zone)
+    utc_iso_format = utc_datetime.strftime('%Y-%m-%dT%H:%M:%SZ')
+
+    return utc_iso_format
+
+
+def convert_utc_timestamp_to_date_and_time(utc_timestamp):
+    utc_zone = dateutil.tz.gettz('UTC')
+    local_zone = dateutil.tz.gettz('America/New_York')
+    utc_datetime = datetime.fromisoformat(utc_timestamp.replace('Z', ''))
+    utc_datetime = utc_datetime.replace(tzinfo=utc_zone)
+    local_datetime = utc_datetime.astimezone(local_zone)
+    date = local_datetime.strftime('%Y-%m-%d')
+    time = local_datetime.strftime('%H:%M')
+
+    return date, time
+
+
+def sync_all_appointments_domain(domain):
+    try:
+        access_token = request_epic_access_token()
+    except Exception:
+        return None
+    # get all patient case ids for domain
+    patient_case_ids = CommCareCase.objects.get_open_case_ids_in_domain_by_type(domain, 'patient')
+    # get all patient cases
+    patient_cases = CommCareCase.objects.get_cases(patient_case_ids)
+
+    # get extension (appointment) cases ids, for all patients
+    appointment_case_ids = CommCareCaseIndex.objects.get_extension_case_ids(
+        domain, patient_case_ids, False, 'appointment')
+    # get appointment cases in commcare
+    appointment_cases = CommCareCase.objects.get_cases(appointment_case_ids)
+
+    # map fhir ids to appointments currently in commcare
+    appointment_map = {
+        appointment_case.get_case_property('fhir_id'): appointment_case for appointment_case in appointment_cases
+    }
+    for patient in patient_cases:
+        patient_helper = CaseHelper(case=patient, domain=domain)
+        patient_fhir_id = patient.get_case_property('fhir_id')
+        if not patient_fhir_id:
+            given = patient.get_case_property('given_name')
+            family = patient.get_case_property('family_name')
+            birthdate = patient.get_case_property('birthdate')
+            if given and family and birthdate:
+                patient_fhir_id = get_patient_fhir_id(given, family, birthdate, access_token)
+
+            if patient_fhir_id is not None:
+                patient_helper.update({'properties': {
+                    'fhir_id': patient_fhir_id,
+                }})
+            else:
+                continue
+
+        epic_appointments_to_add = []
+        epic_appointments_to_update = []
+        # Get all appointments for patient from epic
+        epic_appointment_records = get_epic_appointments_for_patient(patient_fhir_id, access_token)
```

Comment:
Did you look at using [`FHIRImportConfig`](https://github.com/dimagi/commcare-hq/blob/1f7bf89c89c3411774acc6f54465092819ff98ed/corehq/motech/fhir/tasks.py#L85) ([documentation](https://commcare-hq.readthedocs.io/fhir/fhir_import_config.html))?

Would it be a lot of effort to extend it to allow syncing appointments for this project?

* We might need to make some changes to allow for extension cases instead of child cases.
* We would probably need to be smarter regarding how we set case ownership.
* We would either need all cases to store their corresponding FHIR ID in their `external_id` case property, or use case search to fetch them.
* We probably need [to add a (custom?) serializer](https://github.com/dimagi/commcare-hq/blob/fa5179be50c56bc3aa4adefdda15aa4ea3d1dfec/corehq/motech/serializers.py#L2) to handle timestamps for this project.

... and maybe it's a lot more complicated than that.